kernel/src/transaction.rs

```diff
@@ -0,0 +1,124 @@
+use std::sync::Arc;
+
+use crate::actions::visitors::TransactionVisitor;
+use crate::actions::{get_log_schema, TRANSACTION_NAME};
+use crate::snapshot::Snapshot;
+use crate::EngineInterface;
+use crate::{actions::Transaction, DeltaResult};
+
+pub use crate::actions::visitors::TransactionMap;
+pub struct TransactionScanner {
+    snapshot: Arc<Snapshot>,
+}
+
+impl TransactionScanner {
+    pub fn new(snapshot: Arc<Snapshot>) -> Self {
+        TransactionScanner { snapshot }
+    }
+
+    /// Scan the entire log for all application ids but terminate early if a specific application id is provided
+    fn scan_application_transactions(
+        &self,
+        engine: &dyn EngineInterface,
+        application_id: Option<&str>,
+    ) -> DeltaResult<TransactionMap> {
+        let schema = get_log_schema().project(&[TRANSACTION_NAME])?;
+
+        let mut visitor = TransactionVisitor::new(application_id.map(|s| s.to_owned()));
+
+        let iter =
+            self.snapshot
+                .log_segment
+                .replay(engine, schema.clone(), schema.clone(), None)?;
+        for maybe_data in iter {
+            let (txns, _) = maybe_data?;
+            txns.extract(schema.clone(), &mut visitor)?;
+            if application_id.is_some() && visitor.transactions.len() > 0 {
```

Comment:
Clippy error to fix (sorry, I didn't notice before that somebody had to approve pre-merge testing): 
```
error: length comparison to zero
  --> kernel/src/transaction.rs:36:44
   |
36 |             if application_id.is_some() && visitor.transactions.len() > 0 {
   |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: using `!is_empty` is clearer and more explicit: `!visitor.transactions.is_empty()`
```